openedx/core/djangoapps/discussions/docs/decisions/0001-discussion-configuration-api.rst

```diff
@@ -0,0 +1,102 @@
+Discussion provider configuration
+=================================
+
+
+Status
+------
+
+Proposal
+
+
+Context
+-------
+
+As part of the BD-03 initiative (Blended Development, Project 3),
+we want to enable third-party discussion providers to replace the
+default forums experience in edx-platform.
+
+To accomplish this, we should establish a configuration system to
+enable/disable/configure these new discussion-provider plugins;
+the existing/default/legacy forums experience will be included as one of these plugins.
+
+This ADR proposes a new configuration system that allows operators to create
+pre-populated configurations for specific discussions providers,
+which can then be used and customized by course authors/admins.
+
+
+Decision
+--------
+
+We propose to implement this as a Django database-backed model with
+historical records, `LearningContextDiscussionConfig`.
+
+While the underlying model will use `LearningContextKeys` (so that this
+could eventually be extended to cover (sites/programs/etc.), the first
+implementing data structure (`data.py`) can be described as:
+
+    @attr.s(frozen=True)
+    class CourseDiscussionConfigData:
+        course_key = attr.ib(type=CourseKey)
```

Comment:
Should this be a LearningContextKey instead?

Reply:
Let me answer my own question - this is the _course_-specific DiscussionConfigData.  If we had discussions for other types of learning contexts, we'd make other `*Data` classes for those.  Right? :)

Reply:
That begs the question to me about whether we gain anything by having a data class per type of learning context, or whether it should just be `LearningContextDiscussionConfigData` (not that that's not, like, a long name...)

Reply:
Yeah, the more I looked at it, the less helpful the indirection seemed to be.
We can always bake this in later w/ wrappers and subclasses and what not, but until we have those requirements, I'm in favor of keeping it as simple as possible. I've removed the `data.py` abstraction entirely for now; interaction is done directly w/ `django` models.

Reply:
The reason I had for creating that in the first place was that it seemed like the recommended direction based on conventions used by the new learning sequences APIs and its ADR here: https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/content/learning_sequences/docs/decisions/0001-extensions-to-inter-app-apis.rst